source/slang/slang-ir-dce.cpp

```diff
@@ -294,6 +294,235 @@ bool isFirstBlock(IRInst* inst)
+                    if (useField && isFieldUsed(useField))
+                        return true;
+                }
+            }
+        }
+    }
+
+    return false;
+}
+
+bool removeStoresIntoInst(IRInst* ptrInst)
+{
+    bool changed = false;
+
+    List<IRInst*> storesToRemove;
+    for (auto use = ptrInst->firstUse; use; use = use->nextUse)
+    {
+        // If this is a store, remove it.
+        if (auto store = as<IRStore>(use->getUser()))
+        {
+            if (store->getPtr() == ptrInst)
+                storesToRemove.add(store);
+        }
+
+        // If there are any stores into a 'sub-object' of the pointer,
+        // remove them.
+        //
+
+        if (auto subAddr = as<IRFieldAddress>(use->getUser()))
+            changed |= removeStoresIntoInst(subAddr);
+        
+        if (auto subIndex = as<IRGetElementPtr>(use->getUser()))
+            changed |= removeStoresIntoInst(subIndex);
+    }
+
+    for (auto store : storesToRemove)
+    {
+        changed = true;
+        store->removeAndDeallocate();
+    }
+
+    return changed;
+}
+
+bool removeStoresIntoField(IRStructField* field)
+{
+    return removeStoresIntoInst(field->getKey());
+}
+
+bool trimMakeStructOperands(IRStructField* field)
+{
+    // TODO: This can be sped up by considering the full set of fields instead
+    // of one at a time.
+
+    bool changed = false;
+    auto structType = cast<IRStructType>(field->getParent());
+
+    UIndex indexInStruct = 0;
+    for (auto _field : structType->getFields())
+    {
+        if (field == _field)
+            break;
+        indexInStruct++;
+    }
+
+    List<IRInst*> workList;
+    for (auto use = structType->firstUse; use; use = use->nextUse)
+    {
+        if (use->getUser()->getOp() == kIROp_MakeStruct)
+        {
+            workList.add(use->getUser());
+        }
+    }
+
+    IRBuilder builder(field->getModule());
+
+    for (auto makeStruct : workList)
+    {
+        // Make a replacement list of operands.
+        List<IRInst*> newOperands;
+        for (UInt index = 0; index < makeStruct->getOperandCount(); ++index)
+        {
+            if (index == indexInStruct)
+            {
+                // skip..
+                changed = true;
+                continue;
+            }
+            else
+            {
+                newOperands.add(makeStruct->getOperand(index));
+            }
+        }
+
+        builder.setInsertAfter(makeStruct);
+        auto newMakeStruct = builder.emitMakeStruct(makeStruct->getFullType(), newOperands);
+        makeStruct->replaceUsesWith(newMakeStruct);
+    }
+
+    for (auto makeStruct : workList)
+    {
+        makeStruct->removeAndDeallocate();
+    }
+
+    return changed;
+}
+
+bool isStructEmpty(IRType* type)
+{
+    auto structType = as<IRStructType>(type);
+    if (!structType)
+        return false;
+    
+    UCount fieldCount = 0;
+    for (auto field : structType->getFields())
```

Comment:
what if one of the field is also an empty struct?
should we just run type legalization first to cleanup empty struct types?

Reply:
For now, I've changed it to check the fields recursively, but I'll do a follow-up PR to re-arrange the passes a bit. The optimizable type pass probably doesn't need to run for every call to `simplifyIR()`.